### PR TITLE
[OBOD] Bug fix from test_scheduling.py

### LINKED ISSUE #16728: Fix test_scheduling.py::test_many_args failing when it runs at the end of the test
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Seems like when the failure message is published, it invokes failure callback from all subscribed key ids. We should only invoke the failure callback with the specific key id. 

This also cleans up the callback execution model. Now every callback is invoked within the SubscriberChannel. 

TODO
- [x] Waiting for the test_scheduling result. 

## Related issue number

Closes https://github.com/ray-project/ray/issues/16728

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Confirms this fixes the issue. Need to re-write cpp tests again

**Comment by iycheng** [comment]
So the previous ordering issue is related to this one?

**Comment by rkooo567** [comment]
Yes! It turns out that it was a real bug, and I think having other tests probably affected some stress level or sth

File: src/ray/pubsub/subscriber.cc [hunk 1]
```diff
@@ -95,13 +95,17 @@ SubscriptionCallback SubscriberChannel<KeyIdType>::GetCallbackForPubMessage(
...
-    // If the object id is still subscribed, return a subscribe callback.
-    return std::move(maybe_subscription_callback.value());
-  } else {
-    return nullptr;
+  if (!maybe_subscription_callback.has_value()) {
+    return;
   }
+  cum_processed_messages_++;
+  // If the object id is still subscribed, run a callback to the callback io service.
+  std::string channel_name =
```
Comment (iycheng) [inline src/ray/pubsub/subscriber.cc:103]: not sure, but is name() returning rvalue or lvalue? Should it be const str& ?

**Review by fishbone — COMMENTED** [review]
Overall look good to me. One question about the test, does the test cover the update here? The update of the test looks like an API change?

File: src/ray/pubsub/subscriber.cc (hunk 1 above)
Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:103]: Let me check it out.

**Comment by rkooo567** [comment]
@iycheng Are you talking about the cpp test? This does cover that. I just needed to run asio.poll in order to consume callbacks from them

**Comment by rkooo567** [comment]
@iycheng Ah yeah good point! Let me write unit tests for these things

**Comment by rkooo567** [comment]
Note that the real API change only happened in the internal class (that is not explicitly tested), but the outer class definitely needs to test the new behavior

**Comment by rkooo567** [comment]
All comments addressed

**Comment by ericl** [comment]
Merging since this is causing "object lost" errors in master in user workloads.

**Review by ericl — APPROVED** [review]

**Pull request merged**